### PR TITLE
Restaurant owner can log in

### PR DESCRIPTION
**Pull request by jonas-masharqa**
# Login Functionality
[Link to PT](https://www.pivotaltracker.com/story/show/168829431)

## PR content
In this PR we are adding admin login functionality with the help of the gem Devise.

## This PR contains
- Configurations for admin role using devise
- Edit functionality (admin only)
- Feature tests

## What I have learned
- How to add admin boolean
- How to hide content behind an admin require
- How to authenticate certain actions from a controller with devise through admin.
- How to properly git pong through one remote repo to update PR's

## Images
- When non-admin attempts to edit or delete product
Screenshot 2019-10-02 at 17 16 28

---  

-  Admin view
Screenshot 2019-10-02 at 17 16 35

---  

- No-admin view
Screenshot 2019-10-02 at 17 16 42

### CODE REVIEW

**Review by DIzosimov — APPROVED** [review]

**Review by clalil — APPROVED** [review]

**Review by piavW — CHANGES_REQUESTED** [review]
Add pictures from PT and screen-shot please :)

**Review by yasmineezequiel — APPROVED** [review]

**Review by piavW — APPROVED** [review]

File: features/admin_login.feature [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+      | name  | description | price |
+      | pizza | tasty       | 10    |
+    And I visit the landing page
+
+  Scenario: Owner successfully logs in
+    When I click "Login"
+    And I fill in "Email" with "john@doe.com"
+    And I fill in "Password" with "password"
+    And I click "Log in"
+    And I click "Delete"
```
Comment (oliverochman) [inline features/admin_login.feature]: Then??

File: config/routes.rb [hunk 2]
```diff
@@ -1,5 +1,7 @@
 Rails.application.routes.draw do
   devise_for :users
   
+  resources :products
```
Comment (oliverochman) [inline config/routes.rb]: you need to specify which routes that should be available

File: app/views/products/index.html.haml [hunk 3]
```diff
@@ -1,7 +1,10 @@
...
   - @products.each do |product|
-    %li
-      = product.name
-      = product.description
-      = product.price
+    %tr
+      %td= product.name
+      %td= product.description
+      %td= product.price
+    - if current_user.try(:admin?)
```
Comment (oliverochman) [inline app/views/products/index.html.haml]: why cant you use `current_user.admin?` here?

File: app/controllers/products_controller.rb [hunk 4]
```diff
@@ -2,4 +2,33 @@ class ProductsController < ApplicationController
   def index
     @products = Product.all
   end
+
+  before_action :authenticate_admin!, only: [:edit]
```
Comment (oliverochman) [inline app/controllers/products_controller.rb]: put this on line two and write it like this instead:

`before_action :authenticate_admin!, only: [:update, :destroy, :edit]`

File: app/views/products/edit.html.haml [hunk 5]
```diff
@@ -0,0 +1,26 @@
+%h1 Edit product
```
Comment (oliverochman) [inline app/views/products/edit.html.haml]: is editing and deleting products in scope for this feature?

**Review by oliverochman — CHANGES_REQUESTED** [review]

File: app/views/products/edit.html.haml (hunk 5 above)
Comment (clalil) [reply app/views/products/edit.html.haml]: We are only using the edit product part to verify that our admin functionality is working properly (ie. a regular user should not be able to see the edit btn).

File: app/views/products/index.html.haml (hunk 3 above)
Comment (clalil) [reply app/views/products/index.html.haml]: We get undefined method `admin?' for nil:NilClass (ActionView::Template::Error) when we do that. We didn't see why we would need to create a separate Admin model for this particular feature.

File: features/admin_login.feature (hunk 1 above)
Comment (clalil) [reply features/admin_login.feature]: Changed accordingly.

File: app/views/products/edit.html.haml (hunk 5 above)
Comment (oliverochman) [reply app/views/products/edit.html.haml]: Well you can make a button then, not add the whole edit functionality. I think you have a whole story in ur PT board that is about this functionality.

File: config/routes.rb (hunk 2 above)
Comment (clalil) [reply config/routes.rb]: Changes made accordingly.

File: app/controllers/products_controller.rb (hunk 4 above)
Comment (clalil) [reply app/controllers/products_controller.rb]: Changes made accordingly.

File: app/views/products/edit.html.haml (hunk 5 above)
Comment (clalil) [reply app/views/products/edit.html.haml]: Changes made accordingly.

File: app/controllers/products_controller.rb [hunk 6]
```diff
@@ -1,5 +1,34 @@
 class ProductsController < ApplicationController
+
   def index
     @products = Product.all
   end
+
+  before_action :authenticate_admin!, only: [:edit, :update, :destroy]
```
Comment (oliverochman) [inline app/controllers/products_controller.rb]: remove everything in this file that is out of scope for this feature

Comment (DIzosimov) [reply app/controllers/products_controller.rb]: Done :)

File: app/controllers/application_controller.rb [hunk 7]
```diff
@@ -5,6 +5,11 @@ class ApplicationController < ActionController::Base
 
   protected
 
+  def authenticate_admin!
```
Comment (oliverochman) [inline app/controllers/application_controller.rb]: out of scope

**Review by oliverochman — CHANGES_REQUESTED** [review]

File: app/controllers/application_controller.rb (hunk 7 above)
Comment (DIzosimov) [reply app/controllers/application_controller.rb]: Done :)

**Review by DIzosimov — APPROVED** [review]

**Review by yasmineezequiel — APPROVED** [review]
